app/models/keyword_engine.rb

```diff
@@ -0,0 +1,46 @@
+require 'matrix'
+require 'tf-idf-similarity'
+
+class KeywordEngine
+  KEYWORDS_PER_ARTICLE = 4
+  MINIMUM_KEYWORD_LENGTH = 4
+
+  attr_reader :corpus,
+              :model
+
+  def initialize
+    @corpus = create_corpus
+    @model = create_model
+  end
+
+  def create_corpus
+    Article.all.map do |article|
+      base_text = article.keyword_base_text
+      TfIdfSimilarity::Document.new(base_text)
+    end
+  end
+
+  def create_model
+    TfIdfSimilarity::TfIdfModel.new(corpus)
+  end
+
+  def generate_keywords(article)
+    tfidf_by_term = {}
+    document = find_document(article)
+    document.terms.each do |term|
+      tfidf_by_term[term] = model.tfidf(document, term)
+    end
+    sorted_keywords = tfidf_by_term.sort_by{|_,tfidf| -tfidf}
```

Comment:
Space missing after comma.<br>Space missing to the left of {.<br>Space between { and | missing.<br>Space missing inside }.